### ISSUE TITLE
UserSecretsId prevents any deployment of F# ASP.NET Core app to Azure App Service

### ISSUE THREAD

**Issue by cmeeren** [Microsoft/visualfsharp#5549]
When deploying an F# ASP.NET Core app to Azure App Service. the deployment process adds a `<UserSecretsId>` element to the fsproj if it's not already present. As explained [in this article](https://medium.com/@dmytrol/making-asp-net-core-user-secrets-work-in-f-projects-9b04572d81f6), this element causes the build to fail with the following error:

`A function labeled with the 'EntryPointAttribute' attribute must be the last declaration in the last file in the compilation sequence`

The article explains why the error occurs and instructs to fix it by removing the element from the project file and instead adding the user secrets ID in an `AssemblyInfo.fs`. I have tried this and can then build manually, but each time I try to deploy, the deployment process still adds a `<UserSecretsId>` element with a new ID in my project file, causing the build to fail.

In other words, this bug seems to **completely block deployment of F# ASP.NET Core apps to Azure App Service.** It might happen for other Azure deployments too; I haven't tested.

Let me know if I can provide any other relevant info.

**Comment by forki** [comment]
We are putting our asp.net core app into a docker container and run the
container in app service. That works. Maybe a workaround for you

Christer van der Meeren <notifications@github.com> schrieb am Do., 23. Aug.
2018, 10:46:

**Comment by cmeeren** [comment]
Thanks for the suggestion. If I need to use containers, I think I'll rather just wait until our supplier supports AKS (which won't be too long).

**Comment by cartermp** [comment]
cc @vijayrkn @mlorbetske any ideas where this might lie? It looks like it's adding an element to end of the `ItemGroup` in the project file, but this violates order dependence in the F# compiler.

**Comment by vijayrkn** [comment]
+@BillHiebert

**Comment by vijayrkn** [comment]
>When deploying an F# ASP.NET Core app to Azure App Service. the deployment process adds

@cmeeren - How is this entry getting added during deployment?

**Comment by cmeeren** [comment]
@vijayrkn  I think it was first added when I created the publish profile, but if it's missing (e.g. if I delete it) it is re-added to the project file (with a new GUID) each time I click the final "Publish" button.

**Comment by vijayrkn** [comment]
The publish process should not be adding anything to the fsproj. Can you please provide a simple repro project? This will help us diagnose the issue faster.

**Comment by cmeeren** [comment]
I'll try, but how can I do that considering that you'd need a publish profile (I assume with username/password)?

**Comment by vijayrkn** [comment]
you can add the publish profile with no credentials. We will try publishing to a different target

**Comment by cmeeren** [comment]
Great! I'll see what I can do.

**Comment by NinoFloris** [comment]
@vijayrkn this is actually also visible if you just add `<UserSecretsId>Consently</UserSecretsId>
` to an fsproj and run dotnet run

**Comment by cmeeren** [comment]
Here's a repro solution. Just created from template, no changes. Note that the `<UserSecretsId>` element is added once the Publish button is clicked, even before you enter username and password, so you can try to just publish using the existing target and observe the behavior.

[TestApp.zip](https://github.com/Microsoft/visualfsharp/files/2335876/TestApp.zip)

**Comment by vijayrkn** [comment]
Thanks! Taking a look now.

**Comment by vijayrkn** [comment]
@cmeeren - Thank you for the repro project. 

@BillHiebert helped diagnose the issue and we will create a bug in the correct repo. For now, to unblock the scenario, you can delete the UserSecretsId property from the csproj and add the following to the pubxml under Properties\PublishProfiles\profile.pubxml

`    <DisableKeyVaultDuringPublish>true</DisableKeyVaultDuringPublish>`

Once this is added, publish will not add the UserSecretsId  to the csproj.

**Comment by cmeeren** [comment]
Thank you so much for the workaround! Will try it tomorrow.

**Comment by vijayrkn** [comment]
@natemcmaster - This looks like an issue here -  https://github.com/aspnet/DotNetTools/blob/03e73cc8060ad167056d66ee0bf625dfe6943d82/tooling/Microsoft.VisualStudio.SecretManager/ProjectLocalSecretsManager.cs#L79

Requesting for the secretnames is creating a store.  We will have to move this issue to this repo.

**Comment by natemcmaster** [comment]
IMO the better fix would be to get assembly attribute generation working correctly. In the meantime, you can suppress this by adding this to your .fsproj
```xml
<PropertyGroup>
 <GenerateUserSecretsAttribute>false</GenerateUserSecretsAttribute>
</PropertyGroup>
```

**Comment by NinoFloris** [comment]
Guys, not to be annoying but I'd really rather like to see UserSecrets work together with F#

**Comment by natemcmaster** [comment]
Me too. Why doesn't assembly attribute generation work for F#? I don't know enough about the language to understand what this means:

> A function labeled with the 'EntryPointAttribute' attribute must be the last declaration in the last file in the compilation sequence

**Comment by cmeeren** [comment]
For AssemblyInfo generation, see #3113 (if I understand correctly). It's been silent on that front for a while.

**Comment by cartermp** [comment]
As of https://github.com/Microsoft/visualfsharp/pull/4430/files and https://github.com/dotnet/sdk/pull/2023 we should allow it, but only for .NET Core 2.1.

@natemcmaster F# is dependent on declaration and file ordering due to its type system. Nothing can come after a function decorated with `[<EntryPoint>]` due to this.

**Comment by natemcmaster** [comment]
So, if I understand right, we just need to have assembly-attribute generated code in the `<CompileBefore>` item group, not `<Compile>`, right?

**Comment by cartermp** [comment]
I believe so, yes. @KevinRansom could speak more to this, though.

**Comment by KevinRansom** [comment]
@natemcmaster that is correct.

**Comment by natemcmaster** [comment]
Ok, this should be easy enough to fix in ASP.NET Core 2.2: https://github.com/aspnet/Configuration/pull/872

**Comment by natemcmaster** [comment]
@vijayrkn go for it. Can't promise that we'll get to it soon though. It seems pretty low priority. If someone wants to make a PR, I'd welcome it.

**Comment by vijayrkn** [comment]
done - https://github.com/aspnet/DotNetTools/issues/479
